### ISSUE TITLE
Add next js demo. Was: Warning when using Hookstate with Next.JS

### ISSUE THREAD

**Issue by ermaliii** [avkonst/hookstate#223]
Hello,

I am receiving a warning when I use Hookstate with NextJS.

`Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://reactjs.org/link/uselayouteffect-ssr for common fixes.`

I am not able to find any instruction on how to use Hookstate with Next.

Any help is greatly appreciated.

**Comment by avkonst** [comment]
Ohh. This is a bit new and hard one to address. Sometime ago Hookstate used useEffect instead of useLayoutEffect. Unfortunately it was broken under Safari for some states (especially async loaded state), because Safari could dispatch network events or timers (which could mutate the state) before firing useEffect callbacks (which subscribes to mutations events). This lead Hookstate to miss re-rendering in certain cases in Safari.

I will leave it open and think how to address it, but today I am not sure how to remove the working.

But I can assure you that you can ignore the warning, as this "This will lead to a mismatch between the initial, non-hydrated UI and the intended UI" can not be caused by Hookstate useEffect callback.

**Comment by avkonst** [comment]
I am back again. I have tried Hookstate with the latest next version. I can not see this warning. Could you please submit a reproducer (repo or sandbox) for this issue?

**Comment by avkonst** [comment]
Could you please provide more information where you see the warning? I can not reproduce it. If you can, please reopen the ticket.

**Comment by afkcodes** [comment]
> Could you please provide more information where you see the warning? I can not reproduce it. If you can, please reopen the ticket.

actually i was also trying to integrate and saw this showing up in the console don't know from where it came because i have no idea how to first integrate hookstate with next js and i am using next js latest.

**Comment by avkonst** [comment]
Hmm... do you have a reproducer which I could build and run? I tried it and
could not see the warning...

**Comment by afkcodes** [comment]
@avkonst just that im new i might be doing something wrong i was not able to integrate hookstate thats why i am not sure, if we can get a example on how to use hookstate with next js correctly that would be great and if the issue comes up again i would let you know with a reproducer.

**Comment by afkcodes** [comment]
@avkonst any updates please.

**Comment by afkcodes** [comment]
A proper next js implementation example would be great

**Comment by afkcodes** [comment]
`
Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://reactjs.org/link/uselayouteffect-ssr for common fixes.
    at Radio (webpack-internal:///./components/Radio.js:22:3)
    at div
    at div
    at Home (webpack-internal:///./pages/index.js:34:3)
    at EnvironmentProvider (/Users/ashish/tarana-live/node_modules/@chakra-ui/react-env/dist/cjs/env.js:38:24)
    at ColorModeProvider (/Users/ashish/tarana-live/node_modules/@chakra-ui/color-mode/dist/cjs/color-mode-provider.js:51:21)
    at ThemeProvider (/Users/ashish/tarana-live/node_modules/@emotion/react/dist/emotion-element-ce965253.cjs.dev.js:106:21)
    at ThemeProvider (/Users/ashish/tarana-live/node_modules/@chakra-ui/system/dist/cjs/providers.js:24:34)
    at /Users/ashish/tarana-live/node_modules/@chakra-ui/hooks/dist/cjs/use-id.js:22:23
    at /Users/ashish/tarana-live/node_modules/@chakra-ui/hooks/dist/cjs/use-id.js:22:23
    at ChakraProvider (/Users/ashish/tarana-live/node_modules/@chakra-ui/provider/dist/cjs/chakra-provider.js:29:24)
    at Chakra (webpack-internal:///./components/Chakra.js:14:3)
    at MyApp (webpack-internal:///./pages/_app.js:23:3)
    at AppContainer (/Users/ashish/tarana-live/node_modules/next/dist/server/render.js:293:29)
Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://reactjs.org/link/uselayouteffect-ssr for common fixes.
    at Radio (webpack-internal:///./components/Radio.js:22:3)
    at div
    at div
    at Home (webpack-internal:///./pages/index.js:34:3)
    at EnvironmentProvider (/Users/ashish/tarana-live/node_modules/@chakra-ui/react-env/dist/cjs/env.js:38:24)
    at ColorModeProvider (/Users/ashish/tarana-live/node_modules/@chakra-ui/color-mode/dist/cjs/color-mode-provider.js:51:21)
    at ThemeProvider (/Users/ashish/tarana-live/node_modules/@emotion/react/dist/emotion-element-ce965253.cjs.dev.js:106:21)
    at ThemeProvider (/Users/ashish/tarana-live/node_modules/@chakra-ui/system/dist/cjs/providers.js:24:34)
    at /Users/ashish/tarana-live/node_modules/@chakra-ui/hooks/dist/cjs/use-id.js:22:23
    at /Users/ashish/tarana-live/node_modules/@chakra-ui/hooks/dist/cjs/use-id.js:22:23
    at ChakraProvider (/Users/ashish/tarana-live/node_modules/@chakra-ui/provider/dist/cjs/chakra-provider.js:29:24)
    at Chakra (webpack-internal:///./components/Chakra.js:14:3)
    at MyApp (webpack-internal:///./pages/_app.js:23:3)
    at AppContainer (/Users/ashish/tarana-live/node_modules/next/dist/server/render.js:293:29)
`

**Comment by afkcodes** [comment]
Same useLayoutEffect issue that i got today, managed to make hookstate work this seems to pop up everytime

**Comment by avkonst** [comment]
This is all great. Could you please share the reproducer. As I said above I have tried Hookstate with Nextjs and it worked in my basic sample... Maybe I need to build it differently...? I do not know what steps I need to take in order to reproduce the issue.

**Comment by afkcodes** [comment]
Yes this is a reproducer https://github.com/afkcodes/tarana-live/tree/feature/player
it shows up when the page loads.

**Comment by avkonst** [comment]
I can not see where Hookstate is used in your example. It is only imported in index.js but not referenced. if it is not used the bundler should get rid of the unused import?

**Comment by afkcodes** [comment]
the state is in utils/state/global.js
and components/player.js

**Comment by avkonst** [comment]
the first declares the state, the second uses only `set` method. `Set` does not trigger `useEffect` code of the hookstate. Only `useState` does.

**Comment by afkcodes** [comment]
so it seems my implementation is wrong ? can you help me about setting global state from a non react file. as usState can only be used in a react component.

**Comment by avkonst** [comment]
You are setting it correctly. You can set it from anywhere. The ticket was
created about the warning related to use of useEffect in hookstate. I just
do not see where hookstate can cause this warning as it is not actually
used apart set method. I mean when a state is used it is used inside of a
component via usestate. So your reproducer does not reproduce the mentioned
hookstate problem

**Comment by afkcodes** [comment]
okay, is the warning does not show up at your end ? i see it everytime i refresh the page. It shows up in the terminal though.

**Comment by avkonst** [comment]
OK. 1. I do not see the warning on my end. 2. If you see this warning, please remove the import of the state object and all set calls to it and see if it still reproduces the problem.

**Comment by afkcodes** [comment]
removed as you said now its not happening, but literally means i have removed hookstate from the entire project if i remove import & set calls.

**Comment by avkonst** [comment]
very interesting. 
could you please comment out all calls to useEffect within the hookstate code and see if it still reproduces the problem?

**Comment by afkcodes** [comment]
> very interesting.
> could you please comment out all calls to useEffect within the hookstate code and see if it still reproduces the problem?

currently i don't have any useEffect calls in my entire project.

**Comment by avkonst** [comment]
yes, and this is the problem. You also do not call anything from hookstate which call useEffect.... that is why I do not understand how hookstate may cause this warning. That is why I am asking you to comment all useEffect calls WITHIN the Hookstate code. and see it removes the problem

**Comment by afkcodes** [comment]
sorry i didn't get this, what i was saying is i have not used useEffect within the hookstate code. sorry that i didn't get your answer.

**Comment by avkonst** [comment]
I understand that you did not. I can not reproduce it on my machine. I am asking you to make a few changes to hookstate code to locate the problem. Can you update hookstate's code in node_modules folder? Comment out all calls to useEffect in hookstate module?

**Comment by afkcodes** [comment]
yes i will do that . trying now

**Comment by afkcodes** [comment]
```
 React.useLayoutEffect(function () {
        subscribeTarget.subscribe(link);
        return function () {
            link.onUnmount();
            subscribeTarget.unsubscribe(link);
        };
    });
    return link;
```

**Comment by afkcodes** [comment]
```
import React, {useEffect, useLayoutEffect} from 'react';

const canUseDOM = typeof window !== 'undefined';
const useIsomorphicLayoutEffect = canUseDOM ? useLayoutEffect : useEffect;

```

see if this approach of fixing this is fine if you agree i will try to create a PR

**Comment by avkonst** [comment]
This would be the fix I was thinking of. But I do not understand why this warning happens when you this code is never called in your example...? This code is only invoked when you use useState from hookstate. Could you please find out where it is invoked?

**Comment by afkcodes** [comment]
i am using usestate from hookstate in index.js should that be replaced with react usestate ?

**Comment by avkonst** [comment]
well you are not using it, you only importing it...  no you should not replace it if you intended to stay with hookstate

**Comment by afkcodes** [comment]
i will update this and remove unused and report to you

**Comment by avkonst** [comment]
Some magic happens on the import. I am not sure how nextjs detects it... Anyway, If the proposed fix addresses the problem, please submit PR. Thanks

Also, please have a look if actually using the useState function removes the warning. Maybe it is only imported but unused useState triggers the warning?

**Comment by afkcodes** [comment]
update: i am using useState from hookstate at two places at one place i am using it and the other unused i removed it and the warning still shows up.

i will apply the fix and if it fixes the issue will create a PR.

**Comment by afkcodes** [comment]
Created a PR please have a look #234

**Comment by afkcodes** [comment]
Created a PR: #235

**Comment by avkonst** [comment]
fixed in #235

**Comment by afkcodes** [comment]
Hey @avkonst just updated the library in my local but it seems the generated index.js & index.es.js contains the older code. Please tell me if i have done something wrong

**Comment by avkonst** [comment]
released in 3.0.13
